### PR TITLE
Haiku: Initial CoreCLR support

### PR DESCRIPTION
**Pull request by trungnt2910**
This contains a part of the code required to build CoreCLR and get `paltests` to pass on Haiku.

This commit covers `src/coreclr/**`, except the PAL files included in the previous commit.

Part of #55803.

### CODE REVIEW

File: src/coreclr/debug/dbgutil/elfreader.cpp
Comment (janvorli) [inline src/coreclr/debug/dbgutil/elfreader.cpp:329]: I am just curious here - are we disabling this for Haiku because its executable files don't support the stuff the two methods below try to extract or does it just require more work for Haiku?

Comment (trungnt2910) [reply src/coreclr/debug/dbgutil/elfreader.cpp:329]: Don't know about the executable files themselves, but Haiku headers do not have `struct r_debug` and `struct link_map`.

Comment (janvorli) [reply src/coreclr/debug/dbgutil/elfreader.cpp:329]: Ok, got it.

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

**Comment by janvorli** [comment]
@trungnt2910 can you please resolve the conflict so that I can merge the change?

**Comment by trungnt2910** [comment]
Done.

Best to wait for the [Haiku builds CI](https://github.com/trungnt2910/dotnet-builds/actions/runs/11734923750/job/32691756387) to succeed as well before merging though.

**Comment by janvorli** [comment]
@trungnt2910 I am sorry, I've missed you've added note that the Haiku build CI succeeded. Merging now.

**Pull request merged**
